Add RTS#unsafeToFuture to interop module

iravid opened the pull request:
I was looking for something like this for interop, but wasn't sure about the best way to maintain the error information. Looking for some opinions.

This doesn't indicate interruption currently. Not sure what's the best way to express that either.

RaasAhsan commented:
This looks good

We could also think about restricting the conversion functions to only `IO[Nothing, A]`. Its less flexible, but it would force users to handle all checked errors in `IO` (is there any time we wouldn't want to?) and let us pass unchecked errors via `Future`'s underlying `Throwable`. We could also pass a single `Throwable` instead of `CompositeException` if there was only one error.

That would also let us simplify the signature to:
```scala
def unsafeToFuture[A](io: IO[Nothing, A]): Future[A]
```

kaishh commented on interop-future/shared/src/main/scala/scalaz/zio/interop/future.scala:
```diff
@@ -54,6 +54,24 @@ object future {
       }
   }
 
+  case class CompositeException(errors: List[Throwable])
```
I think we should consistently use `FiberFailure` tbh

iravid replied:
Right! Forgot about that. Thanks.

iravid commented:
Sorry I left this hanging. I fixed up the code using the suggestions from the review, split up the code between JVM/JS as RTS is JVM specific for now and tidied up the `futureSpec` tests.

@RaasAhsan / @kaishh, I played around with your suggestions and came around to `unsafeToFuture[E <: Throwable](IO[E, A])` - this will push checked errors, if any, as is into the `Future`, and all the rest as `FiberFailure`. It'll also accept `IO[Nothing, A]`. 

I think this will more closely match user expectations.

jozic commented on interop-future/jvm/src/main/scala/scalaz/zio/interop/FuturePlatformSpecific.scala:
```diff
@@ -0,0 +1,27 @@
+import scala.util.{ Failure, Success }
+
+protected[interop] trait FuturePlatformSpecific {
+  implicit class RTSOps(private val rts: RTS) {
```
wouldn't this create a instance of `RTSOps` on every `unsafeToFuture` call? Could it be made extending `AnyVal`?

iravid replied:
Unfortunately not, because it’s in a trait. Happy to juggle this around if you have an idea.

OTOH, if you’re running unsafeToFuture, you already are allocating monstrously... :-)

kaishh replied:
You could split implicit class into AnyVal in object + implicit def in trait if you wanted to.

iravid replied:
Ahh, thanks!

iravid commented:
Unfortunately not, because it’s in a trait. Happy to juggle this around if you have an idea.

OTOH, if you’re running unsafeToFuture, you already are allocating monstrously... :-)
On 6 Jan 2019, 22:58 +0200, Eugene Platonov <notifications@github.com>, wrote:
> @jozic commented on this pull request.
> In interop-future/jvm/src/main/scala/scalaz/zio/interop/FuturePlatformSpecific.scala:
> > @@ -0,0 +1,27 @@
> +package scalaz.zio.interop
> +
> +import scalaz.zio.{ ExitResult, FiberFailure, IO, RTS }
> +import scala.concurrent.{ Future, Promise }
> +import scala.util.{ Failure, Success }
> +
> +protected[interop] trait FuturePlatformSpecific {
> +  implicit class RTSOps(private val rts: RTS) {
> wouldn't this create a instance of RTSOps on every unsafeToFuture call? Could it be made extending AnyVal?
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

iravid commented:
Anyone got a few moments for a last look? @RaasAhsan @kaishh @regiskuckaertz @jdegoes

iravid commented:
Thanks @regiskuckaertz. It is definitely syntactic sugar for that, but keeping it separate allowed the class to `extend AnyVal` as the enrichment is currently in a platform-specific trait.
